soco/zonegroupstate.py

```diff
@@ -0,0 +1,267 @@
+            Location="http://192.168.1.102:1400/xml/device_description.xml"
+            MinCompatibleVersion="22.0-00000"
+            SoftwareVersion="24.1-74200"
+            UUID="RINCON_000YYY1400"
+            ZoneName="Kitchen"/>
+      </ZoneGroup>
+    </ZoneGroups>
+    <VanishedDevices/>
+  </ZoneGroupState>
+
+"""
+import logging
+import time
+
+from lxml import etree as LXML
+
+from . import config
+from .groups import ZoneGroup
+
+EVENT_CACHE_TIMEOUT = 60
+POLLING_CACHE_TIMEOUT = 5
+NEVER_TIME = -1200.0
+
+ZGS_ATTRIB_MAPPING = {
+    "BootSeq": "_boot_seqnum",
+    "ChannelMapSet": "_channel_map",
+    "HTSatChanMapSet": "_ht_sat_chan_map",
+    "MicEnabled": "_mic_enabled",
+    "UUID": "_uid",
+    "VoiceConfigState": "_voice_config_state",
+    "ZoneName": "_player_name",
+}
+ZGS_XSLT = """
+<xsl:stylesheet version="1.0" xmlns:xsl="http://www.w3.org/1999/XSL/Transform">
+  <xsl:output method="xml" encoding="UTF-8"/>
+  <xsl:template match="*">
+    <xsl:copy>
+     <xsl:copy-of select="@*"/>
+      <xsl:apply-templates select="*">
+       <xsl:sort select="(@Coordinator | @UUID)"/>
+      </xsl:apply-templates>
+    </xsl:copy>
+  </xsl:template>
+</xsl:stylesheet>
+"""
+ZGS_TRANSFORM = LXML.XSLT(LXML.fromstring(ZGS_XSLT))
+
+_LOG = logging.getLogger(__name__)
+
+
+class ZoneGroupState:
+    """Handles processing and caching of ZoneGroupState payloads.
+
+    Only one ZoneGroupState instance is created per Sonos household.
+    """
+
+    def __init__(self):
+        """Initialize the ZoneGroupState instance."""
+        self.all_zones = set()
+        self.groups = set()
+        self.visible_zones = set()
+
+        self._cache_until = NEVER_TIME
+        self._last_zgs = None
+
+        # Statistics
+        self.total_requests = 0
+        self.processed_count = 0
+
+    def clear_cache(self):
+        """Clear the cache timestamp."""
+        self._cache_until = NEVER_TIME
+
+    def clear_zone_groups(self):
+        """Clear all known group sets."""
+        self.groups.clear()
+        self.all_zones.clear()
+        self.visible_zones.clear()
+
+    def poll(self, soco):
+        """Poll using the provided SoCo instance and process the payload."""
+        # pylint: disable=protected-access
+        if time.monotonic() < self._cache_until:
+            self.total_requests += 1
+            _LOG.debug(
+                "Cache still active (GetZoneGroupState) during poll for %s",
+                soco.ip_address,
+            )
+            return
+
+        if soco._is_satellite:
+            # Satellites can return outdated information, use the parent
+            _LOG.debug(
+                "Poll request on satellite (%s), using parent (%s)",
+                soco.ip_address,
+                soco._satellite_parent.ip_address,
+            )
+            soco = soco._satellite_parent
+
+        zgs = soco.zoneGroupTopology.GetZoneGroupState()["ZoneGroupState"]
+        self.process_payload(zgs, "poll", soco.ip_address)
```

Comment:
Maybe add keywords to the last two argument, to indicate their function. I know that at present the called method is right below, but that might change.

Reply:
Done.